py4DSTEM/process/diffraction/utils.py

```diff
@@ -0,0 +1,100 @@
+# Utility functions for the crystal module of py4DSTEM
+
+import numpy as np
+from dataclasses import dataclass
+
+
+@dataclass
+class Orientation:
+    """
+    A class for storing output orientations, generated by fitting a Crystal 
+    class orientation plan or Bloch wave pattern matching to a PointList.
+    """
+    num_matches: int
+    matrix = None
+    family = None
+    corr = None
+    inds = None
+    mirror = None
+    angles = None
+
+    def __post_init__(self):
+        # initialize empty arrays
+        self.matrix = np.zeros((self.num_matches,3,3))
+        self.family = np.zeros((self.num_matches,3,3))
+        self.corr   = np.zeros((self.num_matches))
+        self.inds   = np.zeros((self.num_matches,2), dtype='int')
+        self.mirror = np.zeros((self.num_matches), dtype='bool')
+        self.angles = np.zeros((self.num_matches,3))
```

Comment:
Did @cophus  or @sezelt check if this is the correct way to initialize? I remember Steve mentioning this may create objects that all point at the same mem id?

Reply:
This is not correct!  I can fix it

Reply:
Indeed it needs to use a regular `__init__`. If you use the `__post_init__`, then you can't initialize it using a regular constructor and passing in arrays (i.e. `orient = Orientation(matrix=my_array)` would not actually copy in `my_array`), and if you use the default value syntax then all instances would share the same array (i.e. `a = Orientation(); b = Orientation(); a.matrix is b.matrix # True`).

Reply:
Two issues:

(1) the type hinting is what @dataclass uses to autogenerate and __init__() function for the underlying class.  So everything after num_matrix is not really doing anything (or, actually, is defining a global variable that exists in the namespace of the Class but not in specific class instances, I believe).  So this needs to be changed to, e.g.

`matrix: np.ndarray = None`

(2) Even if that was done correctly, __post_init__ is overwriting all those values.  So any value a use inputs for `matrix` is being overwritten.

So @cophus - the one question I have is, do you want these initialized by default to `None` or `np.zeros((self.num_matches,3,3))`?  The latter, yes?

Reply:
@bsavitzky the test case you wrote here is not the problem with this syntax. When Python evaluates the class definition, it evaluates the expression in the default assignment, and uses that result for every instantiation of the class. In your example, `Test.a` and `Test.b` are indeed distinct expressions and so yield distinct objects. However, the problem is that _every instance of the dataclass gets the same object_. `Test.a` and `Test.b` are different objects, but every time you create a new `Test`, those will all get _the same `a` and `b` arrays_. Example:
```
In [4]: @dataclass
   ...: class Test:
   ...:     a: np.ndarray = np.zeros((2,2))
   ...:     b: np.ndarray = np.zeros((2,2))
   ...:

In [5]: A = Test()

In [6]: B = Test()

In [7]: A.a is B.a
Out[7]: True

In [8]: A.a[0,0] = 42

In [9]: B.a
Out[9]:
array([[42.,  0.],
       [ 0.,  0.]])
``` 

So the course of action is not the same–to get the behavior that we are expecting you _need_ to use an `__init__`, because it gets evaluated every time the class gets instantiated and thus yields unique objects!

(P.S. I am not actually familiar with the behind-the-scenes mechanics of `dataclass`es, so there may be a way to defer evaluation of the defaults, so that they get reëvaluated every time it gets instantiated. Or it may be the case that using an `__init__` is just how you accomplish this...)

(P.P.S. Python does not do anything with the type hints other than evaluate them to check that they are valid expressions, they can literally be anything that doesn't yield an error. They do not enforce any type discipline and do not effect the resulting objects.)

(P.P.P.S Yes, I did use a New Yorker-style guide diaresis in the first postscript.)

Reply:
Cool!